wagtail/contrib/search_promotions/views.py

```diff
@@ -6,80 +6,74 @@
-        queries = queries.annotate(views=functions.Coalesce(Sum("daily_hits__hits"), 0))
-
-    queries = queries.order_by(ordering)
-
-    # Search
-    is_searching = False
-    query_string = request.GET.get("q", "")
-
-    if query_string:
-        queries = queries.filter(query_string__icontains=query_string)
-        is_searching = True
-
-    # Paginate
-    paginator = Paginator(queries, per_page=20)
-    try:
-        queries = paginator.page(request.GET.get("p", 1))
-    except InvalidPage:
-        raise Http404
-
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/results.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/index.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                ),
-            },
-        )
+class SearchPromotionColumn(RelatedObjectsColumn):
+    cell_template_name = "wagtailsearchpromotions/search_promotion_column.html"
+
+
+class IndexView(generic.IndexView):
+    model = Query
+    template_name = "wagtailsearchpromotions/index.html"
+    results_template_name = "wagtailsearchpromotions/index_results.html"
+    context_object_name = "queries"
+    page_title = gettext_lazy("Search Terms")
+    header_icon = "pick"
+    paginate_by = 20
+    permission_policy = ModelPermissionPolicy(SearchPromotion)
+    index_url_name = "wagtailsearchpromotions:index"
+    index_results_url_name = "wagtailsearchpromotions:index_results"
+    _show_breadcrumbs = True
+    search_fields = ["query_string"]
+    add_url_name = "wagtailsearchpromotions:add"
+    add_item_label = gettext_lazy("Add new promoted result")
+    columns = [
+        TitleColumn(
+            "query_string",
+            classname="query_string",
+            label=gettext_lazy("Search term(s)"),
+            width="40%",
+            url_name="wagtailsearchpromotions:edit",
+            sort_key="query_string",
+        ),
+        SearchPromotionColumn(
+            "editors_picks",
+            label=gettext_lazy("Promoted results"),
+            accessor="editors_picks",
+            width="40%",
+        ),
+        Column(
+            "views",
+            label=gettext_lazy("Views (Past week)"),
+            width="20%",
+            sort_key="views",
+        ),
+    ]
+
+    def get_base_queryset(self):
+        queryset = self.model.objects.filter(editors_picks__isnull=False).distinct()
```

Comment:
Unfortunately there seems to be an issue here, the daily hits will be incorrect if the query has more than one `SearchPromotion`. The `editors_picks__isnull=False` causes a `JOIN` with the `SearchPromotion` table, and the `SUM` is done before the `DISTINCT`, so there will be multiple rows for the same `Query` row with different `SearchPromotion` rows. As a result, the `SUM` will be multiplied by the number of `SearchPromotion`s that the `Query` has.

To verify this, you can try using bakerydemo and adding another promoted result for the `patisserie` query. With each promoted search result added, the views count will be multiplied by the amount of promoted search results. For example, with one result it has 15 views, and if you add another result it will become 30. This wasn't an issue with the original code because it uses the `Query.hits` property, but it does a query for each item in the results (N+1 query problem).

One way to solve this that I found is by using a subquery for doing the `editors_picks__isnull=False` filter, e.g.

```suggestion
        has_promotions = SearchPromotion.objects.values_list("query_id", flat=True)
        queryset = self.model.objects.filter(pk__in=has_promotions)
```

Could you add a test case for this too, i.e. creating a `Query` with multiple `SearchPromotion`s, and ensuring the `views` column has the correct value? If not, I can try adding it as I merge.

Reply:
Chipping in to say Sage and I had a lot of fun discussing this problem and its solutions over a call! Nice brain teaser🧠

Reply:
@Stormheg glad to hear it!✌️